### PR TITLE
Added TagHelper IR support

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue - https://github.com/aspnet/Razor/issues/881

@NTaylorMullen @rynowak 

- Haven't done the PreallocatedTagHelperProperty optimization
- Not sure if I set the `SourceRange` of all the new IR nodes correctly

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorIRLoweringPhase.cs [hunk 1]
```diff
@@ -260,6 +259,151 @@ public override void VisitEndDirectiveBlock(DirectiveChunkGenerator chunkGenerat
...
+                    else
+                    {
+                        var addHtmlAttribute = new AddTagHelperHtmlAttributeIRNode()
+                        {
+                            Name = attribute.Name,
+                            ValueStyle = attribute.ValueStyle
+                        };
+
+                        Builder.Push(addHtmlAttribute);
+                        if (attributeValueNode != null)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorIRLoweringPhase.cs]: if it's `null` just do `Builder.Add`

File: src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorIRLoweringPhase.cs [hunk 2]
```diff
@@ -260,6 +259,151 @@ public override void VisitEndDirectiveBlock(DirectiveChunkGenerator chunkGenerat
...
+                            attributeValueNode.Accept(this);
+                        }
+                        Builder.Pop();
+                    }
+                }
+            }
+
+            private void AddExecuteTagHelpers()
+            {
+                Builder.Add(new ExecuteTagHelpersIRNode());
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorIRLoweringPhase.cs]: I think in the prototype I had this as a marker. We could probably represent this in a better way. Fine for now though; just a design note.

File: src/Microsoft.AspNetCore.Razor.Evolution/Intermediate/AddTagHelperHtmlAttributeIRNode.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    {
+        public override IList<RazorIRNode> Children { get; } = new List<RazorIRNode>();
+
+        public override RazorIRNode Parent { get; set; }
+
+        internal override MappingLocation SourceRange { get; set; }
+
+        public string Name { get; set; }
+
+        public IList<RazorIRNode> ValuePieces { get; set; }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Evolution/Intermediate/AddTagHelperHtmlAttributeIRNode.cs]: Nuke, you're representing the value in Children now.

File: src/Microsoft.AspNetCore.Razor.Evolution/Intermediate/CreateTagHelperIRNode.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        public override IList<RazorIRNode> Children { get; } = EmptyArray;
+
+        public override RazorIRNode Parent { get; set; }
+
+        internal override MappingLocation SourceRange { get; set; }
+
+        public string TagHelperTypeName { get; set; }
+
+        internal TagHelperDescriptor AssociatedDescriptor { get; set; }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Evolution/Intermediate/CreateTagHelperIRNode.cs]: `Descriptor`

File: src/Microsoft.AspNetCore.Razor.Evolution/Intermediate/InitializeTagHelperStructureIRNode.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            if (visitor == null)
+            {
+                throw new ArgumentNullException(nameof(visitor));
+            }
+
+            visitor.VisitInitializeTagHelperStructure(this);
+        }
+
+        public override TResult Accept<TResult>(RazorIRNodeVisitor<TResult> visitor)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Evolution/Intermediate/InitializeTagHelperStructureIRNode.cs]: Side note: Should talk about the value of the OfT visitor with @rynowak tomorrow. Doesn't impact this PR at all; just a reminder for myself 👍

File: src/Microsoft.AspNetCore.Razor.Evolution/Intermediate/SetTagHelperPropertyIRNode.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+        public string TagHelperTypeName { get; set; }
+
+        public string PropertyName { get; set; }
+
+        public string AttributeName { get; set; }
+
+        internal HtmlAttributeValueStyle ValueStyle { get; set; }
+
+        internal TagHelperAttributeDescriptor AssociatedDescriptor { get; set; }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Evolution/Intermediate/SetTagHelperPropertyIRNode.cs]: `Descriptor`

File: test/Microsoft.AspNetCore.Razor.Evolution.Test/IntegrationTests/BasicIntegrationTest.cs [hunk 7]
```diff
@@ -72,5 +73,49 @@ public void BuildEngine_CallProcess()
             Assert.NotNull(document.GetSyntaxTree());
             Assert.NotNull(document.GetIRDocument());
         }
+
+        [Fact]
+        public void SimpleTagHelpers()
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Evolution.Test/IntegrationTests/BasicIntegrationTest.cs]: `TagHelpers` are a pretty massive part of the IR. I feel like they should be in their own integration test class.

File: test/Microsoft.AspNetCore.Razor.Evolution.Test/TestFiles/IntegrationTests/BasicIntegrationTest/SimpleTagHelpers.ir.txt [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    Checksum - 
+    NamespaceDeclaration -  - 
+        UsingStatement -  - System
+        UsingStatement -  - System.Threading.Tasks
+        ClassDeclaration -  -  -  -  - 
+            DeclareTagHelperFields - 
+            RazorMethodDeclaration -  -  -  -  - 
+                HtmlContent - (0:0,0 [0] SimpleTagHelpers.cshtml) - <p>Hola</p>\n
+                TagHelper - 
+                    InitializeTagHelperStructure -
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Evolution.Test/TestFiles/IntegrationTests/BasicIntegrationTest/SimpleTagHelpers.ir.txt]: Update the test node serializer to output more information about the TagHelper bits. For example, this guy the tagmode/tagname.

File: src/Microsoft.AspNetCore.Razor.Evolution/Intermediate/AddTagHelperHtmlAttributeIRNode.cs (hunk 3 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Evolution/Intermediate/AddTagHelperHtmlAttributeIRNode.cs]: Oops. I thought I removed it. #fail

File: src/Microsoft.AspNetCore.Razor.Evolution/Intermediate/InitializeTagHelperStructureIRNode.cs (hunk 5 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Evolution/Intermediate/InitializeTagHelperStructureIRNode.cs]: Talked offline. Roslyn does this and the feature is geared specifically for external users traversing IR trees.

**Comment by ajaybhargavb** [comment]
🆙 📅

File: test/Microsoft.AspNetCore.Razor.Evolution.Test/Intermediate/RazorIRAssert.cs [hunk 9]
```diff
@@ -188,6 +189,77 @@ public static void CSharpExpression(string expected, RazorIRNode node)
...
+                Assert.Equal(valueStyle, tagHelperHtmlAttribute.ValueStyle);
+                Children(tagHelperHtmlAttribute, valueValidators);
+            }
+            catch (XunitException e)
+            {
+                throw new IRAssertException(tagHelperHtmlAttribute, tagHelperHtmlAttribute.Children, e.Message, e);
+            }
+        }
+
+        internal static void TagHelperBoundAttribute(
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Evolution.Test/Intermediate/RazorIRAssert.cs]: `SetTagHelperProperty`

File: test/Microsoft.AspNetCore.Razor.Evolution.Test/Intermediate/DefaultRazorIRLoweringPhaseIntegrationTest.cs [hunk 10]
```diff
@@ -188,9 +191,106 @@ public void Lower_WithUsing()
...
+                n => Using("System", n),
+                n => Using(typeof(Task).Namespace, n),
+                n => Assert.IsType<ClassDeclarationIRNode>(n));
+            var @class = @namespace.Children[2];
+            Children(@class,
+                n => TagHelperFieldDeclaration(n, "InputTagHelper"),
+                n => Assert.IsType<RazorMethodDeclarationIRNode>(n));
+            var method = @class.Children[1];
+            var tagHelperNode = SingleChild<TagHelperIRNode>(method);
+            Children(tagHelperNode,
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Evolution.Test/Intermediate/DefaultRazorIRLoweringPhaseIntegrationTest.cs]: Could probably just make a simple TagHelper verifier that validates the first few children so these are more easily read.

File: test/Microsoft.AspNetCore.Razor.Evolution.Test/Intermediate/DefaultRazorIRLoweringPhaseIntegrationTest.cs [hunk 11]
```diff
@@ -188,9 +191,106 @@ public void Lower_WithUsing()
...
+                    v => CSharpAttributeValue(string.Empty, "Hello", v),
+                    v => LiteralAttributeValue(" ", "World", v)),
+                n => Assert.IsType<ExecuteTagHelpersIRNode>(n));
+        }
+
+        [Fact]
+        public void Lower_TagHelpersWithBoundAttribute()
+        {
+            // Arrange
+            var codeDocument = TestRazorCodeDocument.Create(@"<input bound='foo' />");
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Evolution.Test/Intermediate/DefaultRazorIRLoweringPhaseIntegrationTest.cs]: Don't need `@`

File: test/Microsoft.AspNetCore.Razor.Evolution.Test/IntegrationTests/RazorIRNodeWriter.cs [hunk 12]
```diff
@@ -86,6 +86,31 @@ public override void VisitUsingStatement(UsingStatementIRNode node)
...
+        }
+
+        internal override void VisitCreateTagHelper(CreateTagHelperIRNode node)
+        {
+            WriteContentNode(node, node.TagHelperTypeName);
+        }
+
+        internal override void VisitSetTagHelperProperty(SetTagHelperPropertyIRNode node)
+        {
+            WriteContentNode(node, node.AttributeName, node.PropertyName, string.Format("HtmlAttributeValueStyle.{0}", node.ValueStyle));
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Evolution.Test/IntegrationTests/RazorIRNodeWriter.cs]: `nameof(HtmlAttributeValueStyle)` Same below.

File: test/Microsoft.AspNetCore.Razor.Evolution.Test/IntegrationTests/RazorIRNodeWriter.cs [hunk 13]
```diff
@@ -86,6 +86,31 @@ public override void VisitUsingStatement(UsingStatementIRNode node)
...
         }
 
+        internal override void VisitDeclareTagHelperFields(DeclareTagHelperFieldsIRNode node)
+        {
+            WriteContentNode(node, node.UsedTagHelperTypeNames.ToArray());
+        }
+
+        internal override void VisitInitializeTagHelperStructure(InitializeTagHelperStructureIRNode node)
+        {
+            WriteContentNode(node, node.TagName, string.Format("TagMode.{0}", node.TagMode));
```
Comment (NTaylorMullen) [inline test/Microsoft.AspNetCore.Razor.Evolution.Test/IntegrationTests/RazorIRNodeWriter.cs]: `nameof(TagMode)`

**Pull request merged**